### PR TITLE
AP_OpticalFlow: use I2CDevice interface

### PR DESCRIPTION
**Pull request by ricardotk**
Start using I2CDevice interface in AP_OpticalFlow.

Although this is a trivial patch, it's not tested since I don't have the required hardware, maybe some reviewer could test this one for me :new_moon_with_face:

### CODE REVIEW

File: libraries/AP_HAL/AP_HAL_Boards.h [hunk 1]
```diff
@@ -191,6 +191,8 @@
 #define HAL_OS_SOCKETS 1
 #define HAL_STORAGE_SIZE            16384
 #define HAL_STORAGE_SIZE_AVAILABLE  HAL_STORAGE_SIZE
+#define HAL_OPTFLOW_PX4FLOW_I2C_ADDRESS 0x42    // 7-bit address. 8-bit address is 0x84, range 0x42 - 0x49
+#define HAL_OPTFLOW_PX4FLOW_REGISTER    0x16    // Measure Register 22
```
Comment (lucasdemarchi) [inline libraries/AP_HAL/AP_HAL_Boards.h]: this needs to stay where it was... it's a property of the sensor and doesn't change when you change the board

File: libraries/AP_OpticalFlow/AP_OpticalFlow_Linux.cpp [hunk 2]
```diff
@@ -47,50 +49,42 @@ void AP_OpticalFlow_Linux::init(void)
         return;
     }
 
-    // get pointer to i2c bus semaphore
-    AP_HAL::Semaphore *i2c_sem = hal.i2c->get_semaphore();
-    if (i2c_sem == NULL) {
-        return;
-    }
+    AP_HAL::Semaphore *sem = _dev->get_semaphore();
```
Comment (lucasdemarchi) [inline libraries/AP_OpticalFlow/AP_OpticalFlow_Linux.cpp]: you can just call `_dev->get_semaphore()->{take(...),give()}` where needed

File: libraries/AP_OpticalFlow/AP_OpticalFlow_Linux.cpp [hunk 3]
```diff
@@ -47,50 +49,42 @@ void AP_OpticalFlow_Linux::init(void)
...
 bool AP_OpticalFlow_Linux::request_measurement()
 {
     // send measure request to sensor
-    uint8_t cmd = PX4FLOW_REG;
-    if (hal.i2c->writeRegisters(PX4FLOW_I2C_ADDRESS, cmd, 0, nullptr) != 0) {
+    uint8_t cmd = HAL_OPTFLOW_PX4FLOW_REGISTER;
+    if (!_dev->write_register(cmd, 0)) {
         return false;
     }
     return true;
```
Comment (lucasdemarchi) [inline libraries/AP_OpticalFlow/AP_OpticalFlow_Linux.cpp]: this whole function could be a single statement:

```
return _dev->write_register(HAL_OPTFLOW_PX4FLOW_REGISTER, 0);
```

File: libraries/AP_OpticalFlow/AP_OpticalFlow_Linux.cpp [hunk 4]
```diff
@@ -47,50 +49,42 @@ void AP_OpticalFlow_Linux::init(void)
...
 }
 
 bool AP_OpticalFlow_Linux::read(optical_flow_s* report)
 {
-    // get pointer to i2c bus semaphore
-    AP_HAL::Semaphore *i2c_sem = hal.i2c->get_semaphore();
-    if (i2c_sem == NULL) {
-        return false;
-    }
+    AP_HAL::Semaphore *sem = _dev->get_semaphore();
```
Comment (lucasdemarchi) [inline libraries/AP_OpticalFlow/AP_OpticalFlow_Linux.cpp]: same as above... there are very few places using it, so just call _dev->get_semaphore() in each place

File: libraries/AP_OpticalFlow/AP_OpticalFlow_Linux.cpp [hunk 5]
```diff
@@ -101,21 +95,21 @@ bool AP_OpticalFlow_Linux::read(optical_flow_s* report)
...
 
     // Perform the writing and reading in a single command
-    if (PX4FLOW_REG == 0x00) {
-        if (hal.i2c->readRegisters(PX4FLOW_I2C_ADDRESS, 0, I2C_FRAME_SIZE + I2C_INTEGRAL_FRAME_SIZE, val) != 0) {
+    if (HAL_OPTFLOW_PX4FLOW_REGISTER == 0x00) {
+        if (!_dev->read_registers(0, val, I2C_FRAME_SIZE + I2C_INTEGRAL_FRAME_SIZE)) {
             goto fail_transfer;
         }
         memcpy(&f_integral, &(val[I2C_FRAME_SIZE]), I2C_INTEGRAL_FRAME_SIZE);
     }
```
Comment (lucasdemarchi) [inline libraries/AP_OpticalFlow/AP_OpticalFlow_Linux.cpp]: Humn... not related to this patch, but the condition above seems to be dead code?

File: libraries/AP_OpticalFlow/OpticalFlow.cpp [hunk 6]
```diff
@@ -51,7 +51,7 @@ OpticalFlow::OpticalFlow(AP_AHRS_NavEKF& ahrs) :
       CONFIG_HAL_BOARD_SUBTYPE == HAL_BOARD_SUBTYPE_LINUX_BBBMINI
     backend(new AP_OpticalFlow_Onboard(*this, ahrs)),
 #elif CONFIG_HAL_BOARD == HAL_BOARD_LINUX
-    backend(new AP_OpticalFlow_Linux(*this)),
+    backend(new AP_OpticalFlow_Linux(*this, hal.i2c_mgr->get_device(0, HAL_OPTFLOW_PX4FLOW_I2C_ADDRESS))),
```
Comment (lucasdemarchi) [inline libraries/AP_OpticalFlow/OpticalFlow.cpp]: no boards use bus 0. It's actually 1.  Or add a HAL_OPTFLOW_PX4FLOW_I2C_BUS define

File: libraries/AP_HAL/AP_HAL_Boards.h [hunk 7]
```diff
@@ -191,6 +191,8 @@
 #define HAL_OS_SOCKETS 1
 #define HAL_STORAGE_SIZE            16384
 #define HAL_STORAGE_SIZE_AVAILABLE  HAL_STORAGE_SIZE
+#define HAL_OPTFLOW_PX4FLOW_I2C_ADDRESS 0x42    // 7-bit address. 8-bit address is 0x84, range 0x42 - 0x49
```
Comment (lucasdemarchi) [inline libraries/AP_HAL/AP_HAL_Boards.h]: It would be better do define the address if it's not already defined by board subtypes

**Comment by lucasdemarchi** [comment]
I added some comments and pushed a test branch: https://github.com/lucasdemarchi/ardupilot/commits/review/pr-3863

**Comment by ricardotk** [comment]
Force pushed a new version provided by @lucasdemarchi since the changes were minor

File: libraries/AP_OpticalFlow/AP_OpticalFlow_Linux.cpp (hunk 5 above)
Comment (ricardotk) [inline libraries/AP_OpticalFlow/AP_OpticalFlow_Linux.cpp]: Yes, I have the same opinion, maybe @vmayoral could clarify this one for us.

**Comment by lucasdemarchi** [comment]
@rmackay9 @tridge  do you have PX4Flow to test this patch on? I think just making sure it's returning valid data in a bench test is sufficient.

**Comment by lucasdemarchi** [comment]
Applied, thanks.

**Pull request closed without merging**